### PR TITLE
feat: add awsv4 presign and rds util

### PR DESCRIPTION
**Pull request by windmgc**
This PR adds two features:

- **Performing aws_v4 presign on request**
  The aws_v4 presigning flow has a few differences from the aws_v4 signing flow, this PR is imitating [the presigning flow in AWS Go SDK](https://github.com/aws/aws-sdk-go/blob/c7ddff5bbc5ce2facdb433d814b05b3a46d7da05/aws/signer/v4/v4.go#L311)(I'm not very much familiar with JS syntax but the AWS JS SDK would do a similar thing I guess).

- **Providing RDS utils to build IAM auth token**
  This is a pre-step of implementing the IAM Database authentication feature in the Kong Gateway([KAG-167](https://konghq.atlassian.net/browse/KAG-167). For more details about generating IAM RDS Authentication token, please check https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.Connecting.Go.html
and https://github.com/aws/aws-sdk-go/blob/c7ddff5bbc5ce2facdb433d814b05b3a46d7da05/service/rds/rdsutils/connect.go#L36

~Please note that the PR is still working in progress.~

TODO:
- [x] lots of redundant code between v4 and presign, refactor the signing functions to common utils
- [x] fix the comments and annotations of the functions
- [x] tests

[KAG-167]: https://konghq.atlassian.net/browse/KAG-167?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by windmgc** [comment]
@Tieske Could you please review this? Thanks!

I will continue adding more test cases for presign during the review process

File: spec/02-requests/05-presign_v4_spec.lua [hunk 1]
```diff
@@ -0,0 +1,76 @@
+setmetatable(_G, nil)
+
+-- to get a definitive result
+-- luacheck:ignore
+ngx.time = function()
+  return 1667543171
+end
```
Comment (Tieske) [inline spec/02-requests/05-presign_v4_spec.lua:7]: This changes the global environment, and hence will spill over to other tests.

This should probably be in a `setup` block, with a matching `teardown` block to undo the change

**Review by Tieske — CHANGES_REQUESTED** [review]
had a quick look, but running out of time, will have to check again.

File: spec/02-requests/05-presign_v4_spec.lua (hunk 1 above)
Comment (windmgc) [reply spec/02-requests/05-presign_v4_spec.lua]: fixed

**Comment by Tieske** [comment]
@windmgc I added a commit to update the doc comments to render properly.

please check;
```
make docs
open ./docs/index.html
```

**Comment by windmgc** [comment]
@Tieske  thanks a lot!

**Comment by Tieske** [comment]
@windmgc here's my update of the interface, please review; https://github.com/windmgc/lua-resty-aws/pull/1

**Comment by windmgc** [comment]
> @windmgc here's my update of the interface, please review; [windmgc#1](https://github.com/windmgc/lua-resty-aws/pull/1)

rebase merged to reserve commit msgs

**Review by Tieske — APPROVED** [review]
lgtm
